examples/multigrid/setup.h

```diff
@@ -0,0 +1,755 @@
+  for (CeedInt i=0; i<length; i++) {
+    *maxerror = PetscMax(*maxerror, PetscAbsScalar(e[i]));
+  }
+  CeedVectorRestoreArrayRead(collocated_error, &e);
+  ierr = MPI_Allreduce(MPI_IN_PLACE, &maxerror,
+                       1, MPIU_SCALAR, MPIU_MAX, user->comm); CHKERRQ(ierr);
+  CeedVectorDestroy(&collocated_error);
+
+  PetscFunctionReturn(0);
+}
+
+
+
+// BP Options
+typedef enum {
+  CEED_BP3 = 0, CEED_BP4 = 1, CEED_BP5 = 2, CEED_BP6 = 3
+} bpType;
+static const char *const bpTypes[] = {"bp3","bp4","bp5","bp6",
+                                      "bpType","CEED_BP",0};
+
+// BP specific data
+typedef struct {
+  CeedInt vscale, qdatasize, qextra;
+  CeedQFunctionUser setupgeo, setuprhs, apply, error, ident;
+  const char setupgeofname[PETSC_MAX_PATH_LEN], setuprhsfname[PETSC_MAX_PATH_LEN],
+             applyfname[PETSC_MAX_PATH_LEN], errorfname[PETSC_MAX_PATH_LEN],
+             identfname[PETSC_MAX_PATH_LEN];
+  CeedEvalMode inmode, outmode;
+} bpData;
+
+bpData bpOptions[4] = {
+  [CEED_BP3] = {
+    .vscale = 1,
+    .qdatasize = 6,
+    .qextra = 2,
+    .setupgeo = SetupDiffGeo,
+    .setuprhs = SetupDiffRhs,
+    .apply = Diff,
+    .error = Error,
+    .ident = IdentQF,
+    .setupgeofname = PATH(bp3.h:SetupDiffGeo),
+    .setuprhsfname = PATH(bp3.h:SetupDiffRhs),
+    .applyfname = PATH(bp3.h:Diff),
+    .errorfname = PATH(common.h:Error),
+    .identfname = PATH(common.h:IdentQF),
+    .inmode = CEED_EVAL_GRAD,
+    .outmode = CEED_EVAL_GRAD},
+  [CEED_BP4] = {
+    .vscale = 3,
+    .qdatasize = 6,
+    .qextra = 2,
+    .setupgeo = SetupDiffGeo,
+    .setuprhs = SetupDiffRhs3,
+    .apply = Diff3,
+    .error = Error3,
+    .ident = IdentQF3,
+    .setupgeofname = PATH(bp3.h:SetupDiffGeo),
+    .setuprhsfname = PATH(bp4.h:SetupDiffRhs3),
+    .applyfname = PATH(bp4.h:Diff),
+    .errorfname = PATH(common.h:Error3),
+    .identfname = PATH(common.h:IdentQF3),
+    .inmode = CEED_EVAL_GRAD,
+    .outmode = CEED_EVAL_GRAD},
+  [CEED_BP5] = {
+    .vscale = 1,
+    .qdatasize = 6,
+    .qextra = 1,
+    .setupgeo = SetupDiffGeo,
+    .setuprhs = SetupDiffRhs,
+    .apply = Diff,
+    .error = Error,
+    .ident = IdentQF,
+    .setupgeofname = PATH(bp3.h:SetupDiffGeo),
+    .setuprhsfname = PATH(bp3.h:SetupDiffRhs),
+    .applyfname = PATH(bp3.h:Diff),
+    .errorfname = PATH(common.h:Error),
+    .identfname = PATH(common.h:IdentQF),
+    .inmode = CEED_EVAL_GRAD,
+    .outmode = CEED_EVAL_GRAD},
+  [CEED_BP6] = {
+    .vscale = 3,
+    .qdatasize = 6,
+    .qextra = 1,
+    .setupgeo = SetupDiffGeo,
+    .setuprhs = SetupDiffRhs3,
+    .apply = Diff3,
+    .error = Error3,
+    .ident = IdentQF3,
+    .setupgeofname = PATH(bp3.h:SetupDiffGeo),
+    .setuprhsfname = PATH(bp4.h:SetupDiffRhs3),
+    .applyfname = PATH(bp4.h:Diff),
+    .errorfname = PATH(common.h:Error3),
+    .identfname = PATH(common.h:IdentQF3),
+    .inmode = CEED_EVAL_GRAD,
+    .outmode = CEED_EVAL_GRAD}
+};
+
+// PCMG print arrays
```

Comment:
This is a minor nuisance that's different from `KSPType`. Values are an enum, not strings, for `PCMGType` and `PCMGCycleType`. Worth making a PR to PETSc?

Reply:
Use `PCMGTypes[PC_MG_MULTIPLICATIVE]`

Reply:
(I.e., these arrays are already defined in PETSc, just with a slightly different name than you used here.)

Reply:
Thanks. It is different than the setup for KSPType, which is directly a string.